### PR TITLE
bpo-35028: cgi: Fix max_num_fields off by one error

### PR DESCRIPTION
**Pull request by matthewbelisle-wf**
https://bugs.python.org/issue35028

### CODE REVIEW

**Comment by matthewbelisle-wf** [comment]
@ambv @methane Here is the PR I mentioned in https://github.com/python/cpython/pull/9973 for fixing the off by one error. I also improved the unit tests to make them a little clearer.

**Comment by matthewbelisle-wf** [comment]
I figure this shouldn't need a merge entry since there hasn't been a release since https://github.com/python/cpython/pull/9660. If it does let me know and I'll add one.

File: Lib/cgi.py [hunk 1]
```diff
@@ -618,6 +618,12 @@ def read_multi(self, environ, keep_blank_values, strict_parsing):
             first_line = self.fp.readline()
             self.bytes_read += len(first_line)
 
+        # Propagate max_num_fields into the sub class appropriately
+        max_num_fields = self.max_num_fields
+        sub_max_num_fields = self.max_num_fields
+        if max_num_fields is not None:
+            sub_max_num_fields -= len(self.list)
```
Comment (methane) [inline Lib/cgi.py]: I'm not so happy for maintain two local variables.
How about keep only max_num_fields?

```py
            max_num_fields = self.max_num_fields
            if max_num_fields is not None:
                max_num_fields -= len(self.list)
```

Comment (methane) [reply Lib/cgi.py]: Oh, github removed my one comment.

```py
[snip]
part = klass(..., max_num_fields)
[snip]
if max_num_fields is not None:
    max_num_fields -= 1
    if part.list:
        max_num_fields -= len(part.list)
    if max_num_fields < 0:
        raise ValueError('Max number of fields exceeded')
```

Comment (matthewbelisle-wf) [reply Lib/cgi.py]: Good idea, that is simpler. I fixed that in a7203c8.

**Review by methane — APPROVED** [review]

**Pull request merged**